[com_fields] Fetch all custom field values in one query

laoneo opened the pull request:
Pull Request for Issue #14473.

### Summary of Changes
Right now when the values of the fields are fetched, for every field a query is done. This pr changes the behavior to one query for all fields.

Integer indexes are faster to lookup than string indexes, the context column got removed from the field_values table, we gain a bit of performance here as well.

Ping @Bakual and @mbabker for review.
Ping  @alikon and @csthomas did I change the installer and upgrade scripts correctly?

### Testing Instructions
- Create some fields.
- Edit an article.
- Set some values for the fields.
- Save the article.
- View the article on the front.


### Expected result
All of these should work as before, but with less queries. When the debugging is on, then you should see less queries logged.


### Actual result
It works, but with more queries.


### Documentation Changes Required
None.

laoneo committed:
Add function to fetch all fields in one shot

laoneo committed:
Remove the context from the field_values table as it is redundant

laoneo committed:
Fix notices errors

laoneo committed:
Cosmetic

laoneo committed:
Remove the context from the upgrade scripts

laoneo committed:
Remove notice

zero-24 commented on administrator/components/com_fields/models/field.php:
```diff
@@ -581,10 +615,16 @@ public function getFieldValue($fieldId, $context, $itemId)
+		$fieldsQuery = $this->getDbo()->getQuery(true);
+		$fieldsQuery->select($fieldsQuery->qn('id'))
+			->from($fieldsQuery->qn('#__fields'))
+			->where($fieldsQuery->qn('context') . ' = ' . $fieldsQuery->q($context));
```
This looks wrong to me as you have removed context from the db table ;)

laoneo replied:
That query is against #__fields and not #__fields_values. Is correct.

zero-24 replied:
Ah thanks i got confused. :+1:

laoneo replied:
Do you know if we use in the core a delete with an inner join as subselects are slower? I couldn't find a way to do it.

zero-24 replied:
I don't know that we use it. But that does not mean that we don't use it. :smile:

Bakual replied:
I'm not that concerned about performance for deleting fields. It's not done multiple times per second usually. Much more important that it works reliably 😄

alikon replied:
@laoneo  in com_finder there is something similar  https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_finder/helpers/indexer/taxonomy.php#L345

laoneo replied:
I saw that, but also doesn't use an inner join.

csthomas commented on administrator/components/com_fields/helpers/fields.php:
```diff
@@ -163,12 +167,12 @@ public static function getFields($context, $item = null, $prepareValue = false,
 				}
 
-				if ($field->value === '' || $field->value === null)
+				if (!isset($field->value) || $field->value === '' || $field->value === null)
```
`$field->value === null` this will never happen. `isset()` will check that.

laoneo committed:
No need to double check of null

laoneo commented:
Thanks @csthomas removed the double null check.

csthomas commented on administrator/components/com_admin/sql/updates/mysql/3.7.0-2016-08-29.sql:
```diff
@@ -66,11 +66,9 @@ CREATE TABLE IF NOT EXISTS `#__fields_groups` (
   `field_id` int(10) unsigned NOT NULL,
-  `context` varchar(255) NOT NULL,
   `item_id` varchar(255) NOT NULL COMMENT 'Allow references to items which have strings as ids, eg. none db systems.',
   `value` text NOT NULL DEFAULT '',
```
You are not changed it now, but database field with type TEXT NOT NULL can not have default value on mysql.
https://dev.mysql.com/doc/refman/5.7/en/blob.html

There is a lot of such mistakes in joomla but please do not repeat it.
I see 2 solutions. Allow to NULL or always set default value in php code.

laoneo replied:
That sort of stuff should be done in a separate pr, basically on thing in one pr, otherwise it gets harder to review.

csthomas replied:
Yes, I want only mention about it.

laoneo replied:
No problem.

yvesh commented on administrator/components/com_fields/models/fields.php:
```diff
@@ -109,7 +109,7 @@ protected function getStoreId($id = '')
 		$id .= ':' . $this->getState('filter.state');
 		$id .= ':' . $this->getState('filter.group_id');
-		$id .= ':' . print_r($this->getState('filter.language'), true);
+		$id .= ':' . serialize($this->getState('filter.language'));
```
serialize? :-)

laoneo replied:
Why not, it is an array and serialize is also used for the assigned cats.

joomdonation commented on administrator/components/com_fields/models/field.php:
```diff
@@ -570,6 +564,46 @@ public function getFieldValue($fieldId, $context, $itemId)
+
+		if (!key_exists($key, $this->valueCache))
+		{
+			$this->valueCache[$key] = null;
```
If I read the code correctly, this line is not needed. You set **$this->valueCache[$key] = $data;**  later already

laoneo replied:
True

joomdonation commented on administrator/components/com_fields/models/field.php:
```diff
@@ -570,6 +564,46 @@ public function getFieldValue($fieldId, $context, $itemId)
+
+			$query->select(array($query->qn('field_id'), $query->qn('value')))
+				->from($query->qn('#__fields_values'))
+				->where($query->qn('field_id') . ' in (' . implode(',', ArrayHelper::toInteger($fieldIds)) . ')')
```
Should be **IN** instead of **in**.  Am I correct that we always use upper case string for SQL keywords?

Bakual commented:
After thinking about this a bit, the only question I have is if we really need both the "getFieldValue" and "getFieldValues" methods. They do quite the same just instead of using a "->where field_id IN (...)" instead of "->where field_id = ...".
That could also be done using the same method with a conditional based on the field_id parameter (multiple ids vs single integer).

mbabker commented:
Please don't combine the two.  That violates SRP.

Bakual commented:
Stupid question, what is SRP?

rdeutz commented:
single responsibility prinzip

Bakual commented:
Makes sense :smile: 

So better to have a method to fetch the value(s) for a single field and one for fetching values for multiple fields allthough they do basically the same thing?
If that's preferred, then it is fine as it is. It was just something to consider :+1:

mbabker commented:
Right.  Basically you want your methods to be responsible for one task and one task only.  Fetching the data for a single field and fetching the data for multiple fields are two different tasks, will presumably have two different return types (one might be a value object and the other may be a collection of value objects), and may be post-processed by two different code paths (even if you're doing a foreach loop and it just calls the path for the single item, it still has to have that extra logic).  So it's better from a structural standpoint to separate those out.

laoneo committed:
Remove not needed code

laoneo committed:
getValue of a single field is using the getValues function

laoneo commented:
Changed the getValue function to use getValues and removed the not needed code as pointed out by @joomdonation. Additionally added some inline comments to explain what the function is doing.

rdeutz commented:
@laoneo when you on it could you fix the code style issues

laoneo commented:
@rdeutz which ones? Travis is running trough.

laoneo commented:
Ah sorry, didn't realize that drone is now doing the CS stuff. Give me a minute.

laoneo committed:
CS

laoneo commented:
Can drone be restarted somehow?

zero-24 commented:
done @laoneo

laoneo commented:
@rdeutz all good now.

laoneo commented:
Should we also add here a release blocker label as it reduces the amount of SQL queries?

mbabker commented:
Added due to the major performance issues this PR aims to address.

joomdonation commented:
I have tested this item :red_circle: unsuccessfully on c2af37bbbb95629f2ad93756032d9008511a3be1<br /><br />There are SQL error causes by this PR:

1. I created two text fields

2. I created to articles

3. I created a menu item to link to category blog layout menu option

Access to that menu item, see this error

1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ') AND `item_id` = '2'' at line 3

The actual SQL command is 

SELECT `field_id`,`value` FROM `jos_fields_values` WHERE `field_id` IN () AND `item_id` = '2'

Look like in some condition, field_id doesn't have any data for some reasons<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14558">issues.joomla.org/tracker/joomla-cms/14558</a>.</sub>

joomdonation commented:
BTW, in my test, look like the custom fields are not being displayed for Category Blog menu option. So in that case, could we stop querying the database to get data? (I haven't tested this feature much so far, so maybe I just mis-understand how it works)

mbabker commented:
I don't think I would hardcode the plugin context to exclude category views.  In theory someone could do a category layout using fields if desired.

joomdonation commented:
OK. That makes sense (although with default usage like this, it is a waste of resource to query data without using it - It might has performance impact).  Also, one thing about category view which I cannot test right now because above error. Before patch, for each item, an SQL query is is generated to get data for a field, so with 2 articles, 2 fields, 4 query total. I guess if this patch works, there will only one query. Do we want to merge it to one query to get data for all custom fields of all articles (not sure if it is possible)

Bakual commented:
> Do we want to merge it to one query to get data for all custom fields of all articles (not sure if it is possible)

Not possible with the current plugin events. You would need a new event for the whole list.

laoneo committed:
Merge branch 'staging' into cf/performance

laoneo committed:
Check if fields ids is empty

laoneo commented:
@joomdonation I'v added a test if the field ids are empty. Now the error should be gone.

joomdonation commented:
I have tested this item :white_check_mark: successfully on 23fac400b4f87843164d68bbb9cb8782c0b06c78<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14558">issues.joomla.org/tracker/joomla-cms/14558</a>.</sub>

joomdonation commented:
OK. It works now. Unrelated to this PR,  on category blog page, still one query per article if you want to improve (in case it can be improved)

laoneo commented:
Open an issue for it. But as @Bakual said, it needs a new event for it, probably not a bad idea, but I'm not sure if it is doable in the 3 series.

klas commented:
I have tested this item :white_check_mark: successfully on 23fac400b4f87843164d68bbb9cb8782c0b06c78<br /><br />with 100 items in list number of queries is reduced from 1000 to 100, there is also jsut a single query on single item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14558">issues.joomla.org/tracker/joomla-cms/14558</a>.</sub>

klas commented:
With 100 items in list view and 10 fields, queries are reduced from 1000 to 100 (one per item). This is much better but still not quite there (not the fault of this PR), we have the same problem with tags.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14558">issues.joomla.org/tracker/joomla-cms/14558</a>.</sub>

klas commented:
This is much better but still not quite there (not the fault of this PR), we have the same problem with tags.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14558">issues.joomla.org/tracker/joomla-cms/14558</a>.</sub>

laoneo commented:
Guess RTC?

sbouey commented:
work now with Falang

PhilETaylor commented:
Please all test again - I think this broke something in the current staging 

https://github.com/joomla/joomla-cms/issues/14932
